src/chromium/crDevTools.ts

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as fs from 'fs';
+import * as util from 'util';
+import { CRSession } from './crConnection';
+
+const kBindingName = '__pw_devtools__';
+
+// This method intercepts preferences-related DevTools embedder methods
+// and stores preferences as a json file in the browser installation directory.
+export class CRDevTools {
+  private _preferencesPath: string;
+  private _prefs: any;
+
+  constructor(preferencesPath: string) {
+    this._preferencesPath = preferencesPath;
+  }
+
+  async install(session: CRSession) {
+    session.on('Runtime.bindingCalled', async event => {
+      if (event.name !== kBindingName)
+        return;
+      const parsed = JSON.parse(event.payload);
+      let result = undefined;
+      if (parsed.method === 'getPreferences') {
+        if (this._prefs === undefined) {
+          try {
+            const json = await util.promisify(fs.readFile)(this._preferencesPath, 'utf8');
+            this._prefs = JSON.parse(json);
+          } catch (e) {
+            this._prefs = {};
+          }
+        }
+        result = this._prefs;
+      } else if (parsed.method === 'setPreference') {
+        this._prefs[parsed.params[0]] = parsed.params[1];
+        this._save();
+      } else if (parsed.method === 'removePreference') {
+        delete this._prefs[parsed.params[0]];
+        this._save();
+      } else if (parsed.method === 'clearPreferences') {
+        this._prefs = {};
+        this._save();
```

Comment:
can we serialize these saves? I'm afraid they'll race against each other.

Reply:
Done.